docs/topics/compatibility-guides/compatibility-guide-22.md

```diff
@@ -0,0 +1,574 @@
+>
+> **Deprecation cycle**:
+>
+> - 1.8.20: report a warning when using `AbstractDoubleTimeSource`
+> - 2.2.0: raise the warning to an error
+
+## Tools
+
+### Correct `setSource()` function in `KotlinCompileTool` to replace sources
+
+> **Issue**: [KT-59632](https://youtrack.jetbrains.com/issue/KT-59632)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: behavioral
+>
+> **Short summary**: The [`setSource()`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-kotlin-compile-tool/set-source.html#) function in [`KotlinCompileTool`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-kotlin-compile-tool/#) now replaces configured sources instead of adding to them.
+> If you want to add sources without replacing existing ones, use the [`source()`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-kotlin-compile-tool/source.html#) function.
+>
+> **Deprecation cycle**:
+>
+> - 2.2.0: enable new behavior
+
+### Deprecate `KotlinCompilationOutput#resourcesDirProvider` property
+
+> **Issue**: [KT-70620](https://youtrack.jetbrains.com/issue/KT-70620)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: The `KotlinCompilationOutput#resourcesDirProvider` property is deprecated.
+> Use `KotlinSourceSet.resources` in your Gradle build script instead to add additional resource directories.
+>
+> **Deprecation cycle**:
+>
+> - 2.1.0: `KotlinCompilationOutput#resourcesDirProvider` is deprecated with a warning
+> - 2.2.0: raise the warning to an error
+
+### Deprecate `BaseKapt.annotationProcessorOptionProviders` property
+
+> **Issue**: [KT-58009](https://youtrack.jetbrains.com/issue/KT-58009)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: The [`BaseKapt.annotationProcessorOptionProviders`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-base-kapt/annotation-processor-option-providers.html#) property is deprecated in favor of 
+> `BaseKapt.annotationProcessorOptionsProviders`, which accepts a `ListProperty<CommandLineArgumentProvider>` instead of a `MutableList<Any>`.
+> This clearly defines the expected element type and prevents runtime failures caused by adding incorrect elements, such as nested lists.
+> If your current code adds a list as a single element, replace the `add()` function with the `addAll()` function.
+>
+> **Deprecation cycle**:
+>
+> - 2.2.0: enforce new type in the API
+
+### Deprecate `kotlin-android-extensions` plugin
+
+> **Issue**: [KT-72341](https://youtrack.jetbrains.com/issue/KT-72341/)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: The `kotlin-android-extensions` plugin is deprecated. Use a separate plugin, `kotlin-parcelize`,
```

Comment:
Add link to the parcelize plugin? https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.parcelize